### PR TITLE
WIP isolate convergence warning in LogReg

### LINKED ISSUE #215: warnings in Logisticregression with prox newton
Running pytest:
```python
celer/tests/test_logreg.py::test_LogisticRegression[True]
  /home/mathurin/workspace/celer/celer/homotopy.py:311: ConvergenceWarning: Objective did not converge: duality gap: 0.841892524980679, tolerance: 0.005545177444479563. Increasing `tol` may make the solver faster without affecting the results much. 
  Fitting data with very small alpha causes precision issues.
    sol = newton_celer(

celer/tests/test_logreg.py::test_LogisticRegression[True]
  /home/mathurin/workspace/celer/celer/homotopy.py:311: ConvergenceWarning: Objective did not converge: duality gap: 23.09072008747797, tolerance: 0.006931471805599453. Increasing `tol` may make the solver faster without affecting the results much. 
  Fitting data with very small alpha causes precision issues.
    sol = newton_celer(

celer/tests/test_logreg.py::test_LogisticRegression[True]
  /home/mathurin/workspace/celer/celer/homotopy.py:311: ConvergenceWarning: Objective did not converge: duality gap: 2.1031969926275593, tolerance: 0.006931471805599453. Increasing `tol` may make the solver faster without affecting the results much. 
  Fitting data with very small alpha causes precision issues.
    sol = newton_celer(
```

### PR DESCRIPTION
**Pull request by Badr-MOUFAD**
fixes #215

### CODE REVIEW

File: celer/tests/isolate_conv_warn.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+from sklearn.utils.estimator_checks import check_estimator
+
+from celer.dropin_sklearn import LogisticRegression
+from celer.utils.testing import build_dataset
+
+import warnings
+warnings.filterwarnings("error")
+
+
+def isolate_logreg_warning(sparse_X):
```
Comment (mathurinm) [inline celer/tests/isolate_conv_warn.py:13]: this design could be simpler. Can you not just put a breakpoint at the convergence warning in homotopy.py, then run the test and see when it breaks, then dump X, y and alpha manually ?

File: celer/tests/isolate_conv_warn.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+from celer.dropin_sklearn import LogisticRegression
+
+
+def reproduce_convergence_warning():
+    # params
+    n_samples, n_features = 100, 2
+    mu, std = 100, 1
+    min_bound, max_bound = 0, 2
+    tol = 1e-4
+    C = 1.
```
Comment (Badr-MOUFAD) [inline celer/tests/isolate_conv_warn.py:25]: For ``C >= 1e4 / alpha_max`` the ConvergenceWarning disapears.

Comment (mathurinm) [reply celer/tests/isolate_conv_warn.py]: Is this the C, X and y that are used in sklearn test ?

**Comment by mathurinm** [comment]
To clarify: can you get **the exact** `X, y` and regularization strength that cause these convergence warnings output by pytest :
```
celer/tests/test_logreg.py::test_LogisticRegression[True]
  /home/mathurin/workspace/celer/celer/homotopy.py:311: ConvergenceWarning: Objective did not converge: duality gap: 0.841892524980679, tolerance: 0.005545177444479563. Increasing `tol` may make the solver faster without affecting the results much. 
  Fitting data with very small alpha causes precision issues.
    sol = newton_celer(

celer/tests/test_logreg.py::test_LogisticRegression[True]
  /home/mathurin/workspace/celer/celer/homotopy.py:311: ConvergenceWarning: Objective did not converge: duality gap: 23.09072008747797, tolerance: 0.006931471805599453. Increasing `tol` may make the solver faster without affecting the results much. 
  Fitting data with very small alpha causes precision issues.
    sol = newton_celer(

celer/tests/test_logreg.py::test_LogisticRegression[True]
  /home/mathurin/workspace/celer/celer/homotopy.py:311: ConvergenceWarning: Objective did not converge: duality gap: 2.1031969926275593, tolerance: 0.006931471805599453.
 ```
 
 The goal is not just to produce a ConvergenceWarning, this is easy (just put a hard problem with a low regularization, lots of features and few iterations).
 
 The goal is to understand why the solver fails **in the setup of the test**. So the first step is to get the exact problem causing the warning (not to generate random X and y ad to try to get a ConvergenceWarning on it).
 Is it clearer now ?

File: celer/tests/inspect_conv_warn.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+"""Check whether ``C`` is the only reason behind the warning.
```
Comment (mathurinm) [inline celer/tests/inspect_conv_warn.py:1]: avoid having two files with lots of duplicated codes, a single script is enough

File: celer/tests/conv_warning/isolate_conv_warn.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+from celer.dropin_sklearn import LogisticRegression
+from celer.tests.conv_warning.dumped_data import DICT_DATA
+
+
+for check_name, check_props in DICT_DATA.items():
+    X = check_props['X']
+    y = check_props['y']
+    C = check_props['C']
```
Comment (Badr-MOUFAD) [inline celer/tests/conv_warning/isolate_conv_warn.py:8]: ``C`` is constant for all estimators (view its [def](https://github.com/mathurinm/celer/blob/700c280b69902faa0d5430336aa4501d9f53f264/celer/tests/test_logreg.py#L47))

Comment (mathurinm) [reply celer/tests/conv_warning/isolate_conv_warn.py:8]: Aren't x and y small enough to be hard coded in your script ?

Comment (Badr-MOUFAD) [reply celer/tests/conv_warning/isolate_conv_warn.py:8]: ``X`` shape is ``(100, 2)`` with floating numbers. Looking at ``dumped_data.py``, I think putting ``X`` and ``y`` in the same script would be cumbersome.

**Comment by Badr-MOUFAD** [comment]
@mathurinm, I have two hypotheses regarding the ``ConvergenceWarning``:
- ``alpha`` is too small (small depends on ``X``, ``y`` and ``tol``). Hence, a lower bound should be computed to prevent that.
- ``X`` is [not centered](https://github.com/mathurinm/celer/blob/700c280b69902faa0d5430336aa4501d9f53f264/celer/dropin_sklearn.py#L634). Indeed, the dumped data is drawn from ``normal(100, 1)``.

**Comment by mathurinm** [comment]
How is alpha as a fraction of alpha_max, ie norm(X.T @ y, ord=np.inf) ?
What makes you think centering plays a role here ?

**Comment by Badr-MOUFAD** [comment]
- The order of ``alpha / alpha_max`` is ``1e-5``
- I think centering is not implemented (referring to this [line](https://github.com/mathurinm/celer/blob/700c280b69902faa0d5430336aa4501d9f53f264/celer/dropin_sklearn.py#L634). The dumped data, namely ``X`` is drawn from ``N(100, 1)``. I don't get warnings when I test with ``X`` drawn from ``N(0, 1)``

**Comment by mathurinm** [comment]
The fact that the data is not centered does not mean that the solver should not converge on it when fit_intercept=False. Fitting an intercept or not just means solving one optimization problem or another.

File: celer/tests/conv_warning/isolate_conv_warn.py (hunk 4 above)
Comment (mathurinm) [reply celer/tests/conv_warning/isolate_conv_warn.py:11]: Good point, I did not see the file. You can even save them as npy if they are large.

Start simple, with a single case of failure (rather than 3 : chances are, they fail for the same reason)

File: celer/tests/conv_warning/isolate_conv_warn.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    alpha_max = norm(X.T @ y, ord=np.inf)
+    current_alpha = 1. / C
+    # print(f"ratio current_alpha/alpha_max: {current_alpha / alpha_max}")
+
+    # init with alpha_max instead of current_alpha
+    clf = LogisticRegression(C=C, tol=tol, verbose=0, max_iter=50)
+    clf.fit(X, y)
+
+
+# with a standard normal
```
Comment (Badr-MOUFAD) [inline celer/tests/conv_warning/isolate_conv_warn.py:23]: No warning for the centered dataset.

**Comment by Badr-MOUFAD** [comment]
- primal decreases exponentially in the first iteration and slows down (stagnates) later.
- inverse pattern (slightly) is observed for gap.

Also referring to the plot, primal stagnates whereas gap keeps decreasing. It seems like we are stuck in a degenerate point.
Can't we just break when primal stagnates since our objective is to minimize it?

logs

**Comment by josephsalmon** [comment]
Some people do indeed, but mostly when the pb tends to degenerate towards probabilities 0 or 1:
see glmnet paper: [](https://www.jstatsoft.org/article/download/v033i01/361)
image

I have not heard of tricks around 0.5 0.5 (your case here) though.

**Comment by mathurinm** [comment]
The plot displays incomparable quantities: gap goes to 0 while primal converges to a > 0 value, hence primal may continue decreasing at the same speed as the gap but it's not visible. You need to substract the primal limit to the primal objectives if you want comparable quantities

> It seems like we are stuck at a degenerate point.
Degenerate in which sense ?

Stopping based on the primal does not offer guarantees and that is not the way we have chosen in celer.

File: celer/tests/conv_warning/visualize_logs.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
+import re
```
Comment (mathurinm) [inline celer/tests/conv_warning/visualize_logs.py:1]: this is overcomplicated (+ if the error comes from a wrong gap/primal computation you won't see it), run the solver several times while increasing the maximum number of iterations instead

see in the 2 last cells here for example : https://github.com/gpeyre/numerical-tours/blob/master/python/optim_7_noncvx_pro.ipynb

File: celer/tests/conv_warning/visualize_logs.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+import numpy as np
+import matplotlib.pyplot as plt
+
+
+def parse_logs(filename: str) -> np.ndarray:
```
Comment (mathurinm) [inline celer/tests/conv_warning/visualize_logs.py:6]: the more you use complicated code, the more you risk making mistakes without noticing it. try to keep it as simple as possible

**Comment by Badr-MOUFAD** [comment]
@mathurinm, you are right!
When visualizing primal residuals, we get the (quasi) same pattern.

Anyways, in ``LogisticRegression``, when using celer as solver instead of proximal newton, the fit converges in 2 iterations.
refer to[ ``celer/tests/isolate_conv_warn.py``](https://github.com/mathurinm/celer/pull/225/files#diff-5c1392451eee11d6c4ab2b3e2a23bdfe6c8a84ca937edc977ac04be1daad9681)


primal-residuals

**Comment by mathurinm** [comment]
This figure is mathematically impossible, the duality gap is always greater than the primal suboptimality. Can you look into the large peak for the gap ?

**Comment by josephsalmon** [comment]
It is possible, but the double axis scales tricked you. A good reason to never use it.

I agree the peak is huge and looks strange

**Comment by mathurinm** [comment]
Ah yes thanks Joseph. 
@Badr-MOUFAD the double axis is misleading and there is no reason to use it as the two plotted quantities are directly homogeneous and comparable.

Also beware of the way you compute the primal optimum, since we're looking at a convergence issue the last primal after 100 iterations is not necessarily equal to the optimum up to machine precision

**Comment by Badr-MOUFAD** [comment]
> This figure is mathematically impossible, 
> the duality gap is always greater than the primal suboptimality
> Can you also look into the large peak for the gap around iteration 65?

It's true. The scale of the multiaxis plot misled us. 
With that being said, by fixing that, there is no contradiction. The plot of the dual gap is always above the residuals.

I don't have a comment on the pick around iteration 65. Indeed, it doesn't break the previous rule.

Finally, I am pretty sure that there is no (small) mistake in the implementation of  ``newton_celer``, especially knowing that it was coded by @mathurinm.

mul_axis_scaled_plus__

**Comment by Badr-MOUFAD** [comment]
Referring to my knowledge of deep learning, I think that the slowness of the solver might be due to (some sort of) [vanishing gradient](https://en.wikipedia.org/wiki/Vanishing_gradient_problem). Indeed, we compute the gradient of the sigmoid function with data drawn from ``N(100, 1)``.

File: .gitignore [hunk 8]
```diff
@@ -14,6 +14,8 @@ celer.egg-info
 # build
 build
 
+# virtual environment
+venv
```
Comment (Badr-MOUFAD) [inline .gitignore:18]: I isolated celer python environment. 
The ``9K+`` in the source control is a bit disturbing. 
So I added the ``venv`` to ``.gitignore``.

motivation-behind-add-venv

Comment (mathurinm) [reply .gitignore:18]: We don't want to add one line in gitignore for each configuration of each contributor, so use `.git/info/exclude` instead

File: celer/tests/conv_warning/isolate_conv_warn.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+import matplotlib.pyplot as plt
+
+from celer.dropin_sklearn import LogisticRegression
+from celer.tests.conv_warning.logs.dumped_data import DICT_DATA
+
+import pickle
+
+
+# to be excuted only once
+# before, uncomment results in homotopy.py
```
Comment (mathurinm) [inline celer/tests/conv_warning/isolate_conv_warn.py:13]: beware, this is not a reproducible design

File: celer/tests/conv_warning/reproduce_warning.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+# liblinear does not seem to
+clf_sk = sk_LR(C=C, fit_intercept=False, penalty="l1",
+               solver="liblinear", max_iter=10000, tol=1e-8).fit(X, y)
+
+
+# finally, centering the columns yields convergence in very few iterations
+X_c = X - X.mean(axis=0)
+
+y = 2 * y - 1
```
Comment (Badr-MOUFAD) [inline celer/tests/conv_warning/reproduce_warning.py:36]: This is redundant as mapping y to ``[-1, 1]`` is done inside ``LogisticRegression``.

File: celer/tests/conv_warning/reproduce_warning.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+from celer.tests.conv_warning.logs.dumped_data import DICT_DATA
+
+data = DICT_DATA['check_fit_idempotent']
+X = data["X"]  # data centered around 100
+y = data["y"]
+C = data["C"]
+
+y = 2 * y - 1
+C_min = 2 / np.max(np.abs(X.T @ y))
```
Comment (Badr-MOUFAD) [inline celer/tests/conv_warning/reproduce_warning.py:14]: better use ``np.linag.norm(. , ord=np.inf)``.
It s more legible

File: celer/tests/conv_warning/isolate_conv_warn.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        pickle.dump(dict_gaps, f)
+
+    return dict_gaps()
+
+
+# load
+try:
+    with open(filename, 'rb') as f:
+        dict_gaps = pickle.load(f)
+except FileNotFoundError:
```
Comment (Badr-MOUFAD) [inline celer/tests/conv_warning/isolate_conv_warn.py:54]: This will not work.
I will fix it in the next commit.

File: celer/tests/conv_warning/isolate_conv_warn.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+            X, 2 * y_ind - 1, np.array([current_C]), solver="celer-pn")
+
+        current_alpha = 1 / current_C
+        plot_name = f'{current_alpha/alpha_max:.2e}'
+        dict_gaps[plot_name] = gaps
+    # save logs
+    with open(filename, 'wb') as f:
+        pickle.dump(dict_gaps, f)
+
+    return dict_gaps()
```
[the remark this answers is not in the document]
Comment (mathurinm) [reply celer/tests/conv_warning/isolate_conv_warn.py:47]: fix directly

**Comment by mathurinm** [comment]
Thanks for the reproducing scripts @Badr-MOUFAD 

addressed in #227
